miden-crypto/src/merkle/sparse_path.rs

```diff
@@ -0,0 +1,813 @@
+    /// );
+    /// ```
+    pub fn at_idx(&self, index: usize) -> Option<RpoDigest> {
+        // If this overflows *or* if the depth is zero then the index was out of bounds.
+        let depth = NonZero::new(u8::checked_sub(self.depth(), index as u8)?)?;
+        self.at_depth(depth).ok()
+    }
+
+    // PROVIDERS
+    // ============================================================================================
+
+    /// Constructs a borrowing iterator over the nodes in this path.
+    pub fn iter(&self) -> impl ExactSizeIterator<Item = RpoDigest> {
+        self.into_iter()
+    }
+
+    // PRIVATE HELPERS
+    // ============================================================================================
+
+    const fn bitmask_for_depth(node_depth: NonZero<u8>) -> u64 {
+        // - 1 because paths do not include the root.
+        1 << (node_depth.get() - 1)
+    }
+
+    const fn is_depth_empty(&self, node_depth: NonZero<u8>) -> bool {
+        (self.empty_nodes_mask & Self::bitmask_for_depth(node_depth)) != 0
+    }
+
+    fn get_nonempty_index(&self, node_depth: NonZero<u8>) -> usize {
+        let bit_index = node_depth.get() - 1;
+        let without_shallower = self.empty_nodes_mask >> bit_index;
+        let empty_deeper = without_shallower.count_ones() as usize;
+        // The vec index we would use if we didn't have any empty nodes to account for...
+        let normal_index = (self.depth() - node_depth.get()) as usize;
+        // subtracted by the number of empty nodes that are deeper than us.
+        normal_index - empty_deeper
+    }
+}
+
+// SERIALIZATION
+// ================================================================================================
+
+impl Serializable for SparseMerklePath {
+    fn write_into<W: winter_utils::ByteWriter>(&self, target: &mut W) {
+        target.write_u8(self.depth());
+        target.write_u64(self.empty_nodes_mask);
+        target.write_many(&self.nodes);
+    }
+}
+
+impl Deserializable for SparseMerklePath {
+    fn read_from<R: winter_utils::ByteReader>(
+        source: &mut R,
+    ) -> Result<Self, DeserializationError> {
+        let depth = source.read_u8()?;
+        let empty_nodes_mask = source.read_u64()?;
+        let count = depth as u32 - empty_nodes_mask.count_ones();
+        let nodes = source.read_many::<RpoDigest>(count as usize)?;
+        Ok(Self { empty_nodes_mask, nodes })
+    }
+}
+
+// CONVERSIONS
+// ================================================================================================
+
+impl From<SparseMerklePath> for MerklePath {
+    fn from(sparse_path: SparseMerklePath) -> Self {
+        MerklePath::from_iter(sparse_path)
+    }
+}
+
+/// # Errors
+///
+/// This conversion returns [MerkleError::DepthTooBig] if the path length is greater than
+/// [`SMT_MAX_DEPTH`].
+impl TryFrom<MerklePath> for SparseMerklePath {
+    type Error = MerkleError;
+
+    fn try_from(path: MerklePath) -> Result<Self, MerkleError> {
```

Comment:
Nit: I would add the doc comment at the function level. That way, when I hover over `SparseMerklePath::try_from` in my IDE, I get the docs.